server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -36,417 +40,479 @@
+                    continue;
+                }
 
-            final var assignment = desiredBalance.getAssignment(shardRouting.shardId());
-            if (assignment == null) {
-                // balance is not computed
-                continue;
-            }
+                undesiredAllocations++;
 
-            if (assignment.nodeIds().contains(shardRouting.currentNodeId())) {
-                // shard is already on a desired node
-                continue;
-            }
+                if (allocation.deciders().canRebalance(shardRouting, allocation).type() != Decision.Type.YES) {
+                    // rebalancing disabled for this shard
+                    continue;
+                }
 
-            if (allocation.deciders().canRebalance(shardRouting, allocation).type() != Decision.Type.YES) {
-                // rebalancing disabled for this shard
-                continue;
-            }
+                if (allocation.deciders().canAllocate(shardRouting, allocation).type() != Decision.Type.YES) {
+                    // cannot allocate anywhere, no point in looking for a target node
+                    continue;
+                }
 
-            if (allocation.deciders().canAllocate(shardRouting, allocation).type() != Decision.Type.YES) {
-                // cannot allocate anywhere, no point in looking for a target node
-                continue;
+                final var rebalanceTarget = findRelocationTarget(shardRouting, assignment.nodeIds(), this::decideCanAllocate);
+                if (rebalanceTarget != null) {
+                    routingNodes.relocateOrReinitializeShard(
+                        shardRouting,
+                        rebalanceTarget.getId(),
+                        allocation.clusterInfo().getShardSize(shardRouting, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE),
+                        allocation.changes()
+                    );
+                }
             }
 
-            final var rebalanceTarget = findRelocationTarget(shardRouting, assignment.nodeIds(), this::decideCanAllocate);
-            if (rebalanceTarget != null) {
-                routingNodes.relocateOrReinitializeShard(
-                    shardRouting,
-                    rebalanceTarget.getId(),
-                    allocation.clusterInfo().getShardSize(shardRouting, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE),
-                    allocation.changes()
+            maybeLogUndesiredAllocationsWarning(allAllocations, undesiredAllocations);
+        }
+
+        private void maybeLogUndesiredAllocationsWarning(long allAllocations, long undesiredAllocations) {
+            if (allAllocations > 0 && undesiredAllocations > undesiredAllocationsLogThreshold * allAllocations) {
+                undesiredAllocationLogInterval.maybeExecute(
+                    () -> logger.warn("{}/{} of shards are allocated on undesired nodes", allAllocations, undesiredAllocations)
                 );
             }
         }
```

Comment:
Actual logging here, the rest is wiring changes that could also be reviewed commit-by-commit

Reply:
I also wonder if we should have some lower level limit to avoid logging when relocating the only shard in the cluster.
May be when `undesiredAllocationsLogThreshold * allAllocations >= 1`?